src/mumble/GlobalShortcut_win.h

```diff
@@ -6,135 +6,94 @@
+	void injectRawInputMessage(HRAWINPUT handle);
 
-	/// Handle an incoming Windows mouse message.
-	///
-	/// Returns true if the GlobalShortcut engine signals that the
-	/// button should be suppressed. Returns false otherwise.
-	static bool handleMouseMessage(unsigned int btn, bool down);
+	ButtonInfo buttonInfo(const QVariant &button) override;
+	void run() override;
 
-	virtual bool canSuppress() Q_DECL_OVERRIDE;
-	void run() Q_DECL_OVERRIDE;
-	bool event(QEvent *e) Q_DECL_OVERRIDE;
+	GlobalShortcutWin();
+	~GlobalShortcutWin() override;
 public slots:
+	void deviceRemoved(const HANDLE deviceHandle);
 	void timeTicked();
 
-public:
-	GlobalShortcutWin();
-	~GlobalShortcutWin() Q_DECL_OVERRIDE;
-	void unacquire();
-	ButtonInfo buttonInfo(const QVariant &) Q_DECL_OVERRIDE;
-
-	/// Inject a native Windows keyboard message into GlobalShortcutWin's
-	/// event stream. This method is meant to be called from the main thread
-	/// to pass native Windows keyboard messages to GlobalShortcutWin.
-	///
-	/// @param  msg  The keyboard message to inject into GlobalShortcutWin.
-	///              Must be WM_KEYDOWN, WM_KEYUP, WM_SYSKEYDOWN or WM_SYSKEYUP.
-	///              Otherwise the message will be ignored.
-	///
-	/// @return      Returns true if the GlobalShortcut engine signalled that
-	///              the button should be suppressed. Returns false otherwise.
-	bool injectKeyboardMessage(MSG *msg);
+	void on_hidMessage(const HANDLE deviceHandle, std::vector< char > reports, const uint32_t reportSize);
+	void on_keyboardMessage(const uint16_t flags, uint16_t scanCode, const uint16_t virtualKey);
+	void on_mouseMessage(const uint16_t flags, const uint16_t data);
+signals:
+	void hidMessage(const HANDLE deviceHandle, std::vector< char > reports, const uint32_t reportSize);
+	void keyboardMessage(const uint16_t flags, uint16_t scanCode, const uint16_t virtualKey);
+	void mouseMessage(const uint16_t flags, const uint16_t data);
+
+protected:
+	struct Device {
+		std::string name;
+		std::string prefix;
+		std::vector< uint8_t > data;
+		std::vector< bool > buttons;
+		std::pair< uint16_t, uint16_t > usageRange;
+#ifdef USE_XBOXINPUT
+		bool xinput;
+		Device() : xinput(false){};
+#endif
+	};
 
-	/// Inject a native Windows mouse message into GlobalShortcutWin's
-	/// event stream. This method is meant to be called from the main thread
-	/// to pass native Windows mouse messages to GlobalShortcutWin.
-	///
-	/// @param  msg  The keyboard message to inject into GlobalShortcutWin.
-	///              Must be WM_LBUTTONDOWN, WM_LBUTTONUP, WM_RBUTTONDOWN,
-	///              WM_RBUTTONUP, WM_MBUTTONDOWN, WM_MBUTTONUP, WM_XBUTTONDOWN
-	///              or WM_XBUTTONUP. Otherwise the message will be ignored.
-	///
-	/// @return      Returns true if the GlobalShortcut engine signalled that
-	///              the button should be suppressed. Returns false otherwise.
-	bool injectMouseMessage(MSG *msg);
+	typedef std::unordered_map< HANDLE, Device > Devices;
```

Comment:
I would probably prefer to name this `DeviceMap`

Reply:
Okay.